benchexec: init at 3.21

lorenzleutgeb opened the pull request:
## Description of changes

Packages three related programs, the main one being BenchExec. All three programs are maintained by @PhilippWendler at the [Software and Computational Systems Lab](https://sosy-lab.org/) at Ludwig-Maximilians-Universität München.

### [BenchExec](https://github.com/sosy-lab/BenchExec) (version 3.21)

A framework for reliable benchmarking and resource measurement. See also [doi:10.1007/s10009-017-0469-y](https://doi.org/10.1007/s10009-017-0469-y).

### [CPU Energy Meter](https://github.com/sosy-lab/cpu-energy-meter) (version 1.2)

A tool for measuring energy consumption of Intel CPUs. Requires reading [`msr(4)`](https://man7.org/linux/man-pages/man4/msr.4.html) devices. See also [doi:10.1007/978-3-030-45237-7_8](https://doi.org/10.1007/978-3-030-45237-7_8).

### [PQoS Wrapper](https://gitlab.com/sosy-lab/software/pqos-wrapper)

Wrapper for Intel PQoS for the purpose of using it in BenchExec. Requires reading [`msr(4)`](https://man7.org/linux/man-pages/man4/msr.4.html) devices.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

lorenzleutgeb committed:
cpu-energy-meter: init at 1.2

Co-authored-by: h7x4 <h7x4@nani.wtf>

lorenzleutgeb committed:
pqos-wrapper: init at unstable-2022-01-31

Co-authored-by: h7x4 <h7x4@nani.wtf>

lorenzleutgeb committed:
benchexec: init at 3.21

Co-authored-by: h7x4 <h7x4@nani.wtf>

lorenzleutgeb committed:
nixos/doc/rl-2405: mention BenchExec and friends

PhilippWendler commented:
@globin ping because you may like to see this. And your work is included here, it was released as part of [BenchExec 3.18](https://github.com/sosy-lab/benchexec/releases/tag/3.18).

h7x4 commented on pkgs/by-name/cp/cpu-energy-meter/package.nix:
```diff
@@ -0,0 +1,41 @@
+    hash = "sha256-QW65Z8mRYLHcyLeOtNAHjwPNWAUP214wqIYclK+whFw=";
+  };
+
+  patchPhase = ''
+    substituteInPlace Makefile \
+      --replace "DESTDIR :=" "DESTDIR := $out" \
+      --replace "PREFIX := /usr/local" "PREFIX :="
+  '';
```
[Don't forget the `runHook`s :) (see checklist)](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#new-packages)

```suggestion
  patchPhase = ''
    runHook prePatch
  
    substituteInPlace Makefile \
      --replace "DESTDIR :=" "DESTDIR := $out" \
      --replace "PREFIX := /usr/local" "PREFIX :="
      
    runHook postPatch
  '';
```

h7x4 commented on pkgs/by-name/cp/cpu-energy-meter/package.nix:
```diff
@@ -0,0 +1,41 @@
+  env.NIX_CFLAGS_COMPILE = "-fcommon";
+
+  postInstall = ''
+    install -D --mode=444 --target-directory=$out/etc/udev/rules.d \
+      $src/debian/additional_files/59-msr.rules
```
This could be shortened down if you don't want it taking 2 lines.

```suggestion
    install -Dm444 -t $out/etc/udev/rules.d $src/debian/additional_files/59-msr.rules
```

h7x4 commented on nixos/modules/programs/cpu-energy-meter.nix:
```diff
@@ -0,0 +1,29 @@
+, pkgs
+, ...
+}: {
+  options.programs.cpu-energy-meter = with lib.options; {
+    enable = mkEnableOption "CPU Energy Meter";
+    package = mkPackageOption pkgs "cpu-energy-meter" { };
+  };
```
```suggestion
  options.programs.cpu-energy-meter = {
    enable = lib.mkEnableOption "CPU Energy Meter";
    package = lib.mkPackageOption pkgs "cpu-energy-meter" { };
  };
```

h7x4 commented on nixos/modules/programs/pqos-wrapper.nix:
```diff
@@ -0,0 +1,29 @@
+, pkgs
+, ...
+}: {
+  options.programs.pqos-wrapper = with lib.options; {
+    enable = mkEnableOption "pqos-wrapper";
+    package = mkPackageOption pkgs "pqos-wrapper" { };
+  };
```
```suggestion
  options.programs.pqos-wrapper = {
    enable = lib.mkEnableOption "pqos-wrapper";
    package = lib.mkPackageOption pkgs "pqos-wrapper" { };
  };
```

h7x4 commented on pkgs/by-name/pq/pqos-wrapper/package.nix:
```diff
@@ -0,0 +1,28 @@
+    hash = "sha256-SaYr6lVucpJjVtGgxRbDGYbOoBwdfEDVKtvD+M1L0o4=";
+  };
+
+  makeWrapperArgs = [ "--prefix LD_LIBRARY_PATH : ${lib.getLib intel-cmt-cat}/lib" ];
```
Maybe this would be better?

```suggestion
  makeWrapperArgs = [ "--prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [ intel-cmt-cat ]}" ];
```

h7x4 commented on pkgs/by-name/be/benchexec/package.nix:
```diff
@@ -0,0 +1,51 @@
+, libseccomp
+, nixosTests
+}:
+python3.pkgs.buildPythonApplication rec {
```
I don't think python packages like this one belong to `pkgs/by-name` just yet. It's a challenge for later.

https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md#limitations

lorenzleutgeb replied:
OK, will move them.

h7x4 commented on pkgs/by-name/be/benchexec/package.nix:
```diff
@@ -0,0 +1,51 @@
+      # If we add `cpu-energy-meter` here, BenchExec will instead call an executable
+      # without `CAP_SYS_RAWIO` and fail.
+      #pkgs.cpu-energy-meter
+      libseccomp.lib
```
Is this required both as a propagated build input and as an environment variable? Maybe it's possible to drop one of them?

lorenzleutgeb replied:
I'll test that.

h7x4 reviewed (commented):
Haven't tested or went rigorously through this, these are just some initial comments

lorenzleutgeb commented on nixos/modules/programs/benchexec.nix:
```diff
@@ -0,0 +1,59 @@
+
+  config = lib.mkIf cfg.enable {
+    assertions = map
+      (user: {
+        assertion = config.users.users ? ${user};
+        message = ''
+          The user '${user}' intends to use BenchExec (via `${opt.users}`), but is not configured via `${options.users.users}`.
+        '';
+      })
+      cfg.users;
```
Not supported, since don't know how (it's not possible?) to configure `Delegate=yes` for their units via configuration. Maybe this could be worked around by some magic that runs at system boot, but I do not want to go that far.

lorenzleutgeb replied:
Mutable users are now supported by adding their UID (as an integer) to `programs.benchexec.users`.

lorenzleutgeb commented on nixos/modules/programs/benchexec.nix:
```diff
@@ -0,0 +1,59 @@
+      enableUnifiedCgroupHierarchy = true;
+
+      # See <https://github.com/sosy-lab/benchexec/blob/3.18/doc/INSTALL.md#setting-up-cgroups>.
+      services = builtins.listToAttrs
+        (user: {
+          name = "user@${builtins.toString config.users.users.${user}.uid}";
+          value.serviceConfig.Delegate = "yes";
+        })
+        cfg.users;
+    };
```
The goal is to set `Delegate=yes` for every affected user's [`user@.service`](https://www.freedesktop.org/software/systemd/man/latest/user@.service.html). This is according to [the installation instructions of BenchExec](https://github.com/sosy-lab/benchexec/blob/3.18/doc/INSTALL.md#setting-up-cgroups) with the exception that they suggest to do this for all users of the system. The [maintainer commented that it's a good idea to do it per user if possible](https://github.com/sosy-lab/benchexec/pull/791#issuecomment-1772779159).

Now, there might be better ways to achieve this. Happy to learn.

h7x4 replied:
Right, so if I understood this correctly, it has do to with the security implications of letting any user access `CAP_SYS_RAWIO`?

I guess there's a tough choice to make between choosing users individually while lacking mutable user support and generating individual user specific services, versus saying it's a global thing with easier config and to ask people please not to enable it if they are worried about abuse.

Alternatively, it might be possible to implement both, and switch between them using an option. I'm fine with any of these options to be honest, so I think I'll just let this be for now and wait for more opinions.

lorenzleutgeb replied:
No, it's orthogonal to `CAP_SYS_RAWIO`. `CAP_SYS_RAWIO` is only required to open MSR devices.

This is required because BenchExec sets up control groups for the programs that it benchmarks. And in order for it to be able to do so, delegation of control groups must be enabled.

h7x4 replied:
Another, somewhat jank way of solving this might be to add some sort of preExec script to a global user service, check that you are part of a certain group, or else just shut down the unit. I'm not sure if it would be possible to hack your way around using systemd overrides.

lorenzleutgeb replied:
Sorry, I don't understand. The `preExec` script would have to reconfigure the `user@.service`, which it can't do?!

h7x4 replied:
No, more like the user@.service (which should not be modifiable by the user, because it exists in /etc/systemd or the nix store or something), will have a PreExec that will check which user runs the program. If the user does not exist in the correct group, or for any other reasons don't have correct permissions to run the service, PreExec should shut down or intentionally crash such that Exec never gets run.

I haven't ever done such a thing before though, and there are some overriding features in systemd which might make that idea worthless (if the user could just remove the PreExec). I have no clue whether it's even something worth looking into.

lorenzleutgeb replied:
I thought about this for a bit, and I believe a better solution would be a drop-in override:

```nix
      services = builtins.listToAttrs
        (user: {
          name = "user@${builtins.toString config.users.users.${user}.uid}";
          value = {
            serviceConfig.Delegate = "yes";
            overrideStrategy = "asDropin";
          };
        })
        cfg.users;
```

lorenzleutgeb replied:
Please check the changes and reply. Otherwise, since this has been stalling for 6 months, I'll just mark it as resolved to unblock the PR.

lorenzleutgeb commented:
I pushed a test, but am running into a strange failure. Note that I would like to use `systemd-run` instead of `sudo` to connect to DBus. The command works locally on my machine. I tried modifying it, but am really puzzled, and would appreciate any help fixing this. Thank you!

<details><summary>Log output of failing test.</summary>
<pre>
$ nix log '/nix/store/02bi36g2cjvza4q6x53ill5xj9nkga53-vm-test-run-benchexec.drv^*' | tail -n 48
(finished: waiting for unit multi-user.target, in 7.51 seconds)
benchexec # [    7.257393] systemd[1]: Started systemd-stdio-bridge -punix:path=${XDG_RUNTIME_DIR}/bus.
benchexec # [    7.273501] (o-bridge)[973]: pam_unix(login:session): session opened for user alice(uid=1000) by (uid=0)
benchexec # [    7.289689] systemd[1]: Created slice Slice /user/1000.
benchexec # [    7.290994] systemd[1]: Starting User Runtime Directory /run/user/1000...
benchexec # [    7.304226] systemd[1]: Finished User Runtime Directory /run/user/1000.
benchexec # [    7.308417] systemd[1]: Starting User Manager for UID 1000...
benchexec # [    7.316750] (systemd)[977]: pam_unix(systemd-user:session): session opened for user alice(uid=1000) by (uid=0)
benchexec # [    7.407299] systemd[977]: Queued start job for default target Main User Target.
benchexec # [    7.412111] systemd[977]: Created slice User Application Slice.
benchexec # [    7.412612] systemd[977]: Reached target Paths.
benchexec # [    7.413012] systemd[977]: Reached target Timers.
benchexec # [    7.414486] systemd[977]: Starting D-Bus User Message Bus Socket...
benchexec # [    7.427520] systemd[977]: Listening on D-Bus User Message Bus Socket.
benchexec # [    7.428156] systemd[977]: Reached target Sockets.
benchexec # [    7.428542] systemd[977]: Reached target Basic System.
benchexec # [    7.430077] systemd[1]: Started User Manager for UID 1000.
benchexec # [    7.437569] systemd[1]: Started Session 1 of User alice.
benchexec # [    7.437958] systemd[977]: Starting Run user-specific NixOS activation...
benchexec # [    7.445329] systemd[977]: Created slice User Core Session Slice.
benchexec # [    7.448573] systemd[977]: Starting D-Bus User Message Bus...
benchexec # [    7.456623] systemd[977]: Started D-Bus User Message Bus.
benchexec # [    7.459028] dbus-daemon[988]: [session uid=1000 pid=988] Successfully activated service 'org.freedesktop.systemd1'
benchexec # [    7.463416] systemd[977]: Started /nix/store/gi9hyb9psxfngap6rp59z9avx0bw63gj-benchexec-3.18/bin/runexec --debug --read-only-dir / --hidden-dir /home --dir /tmp /nix/store/0k7la5q51v1rw2id0xfzy9266hk0sghp-test/bin/test.
benchexec # [    7.468638] systemd[977]: Finished Run user-specific NixOS activation.
benchexec # [    7.475580] systemd[977]: Reached target Main User Target.
benchexec # [    7.476543] systemd[977]: Startup finished in 151ms.
benchexec # [    7.831466] systemd[1]: run-u3.service: Deactivated successfully.
cleanup
kill machine (pid 8)
benchexec # qemu-kvm: terminating on signal 15 from pid 6 (/nix/store/ffll6glz3gwx342z0ch8wx30p5cnqz1z-python3-3.11.5/bin/python3.11)
(finished: cleanup, in 0.05 seconds)
Traceback (most recent call last):
  File "/nix/store/n6zqh28rh2r2h42npr5drz616khdrjj2-nixos-test-driver-1.1/bin/.nixos-test-driver-wrapped", line 9, in <module>
    sys.exit(main())
             ^^^^^^
  File "/nix/store/n6zqh28rh2r2h42npr5drz616khdrjj2-nixos-test-driver-1.1/lib/python3.11/site-packages/test_driver/__init__.py", line 117, in main
    driver.run_tests()
  File "/nix/store/n6zqh28rh2r2h42npr5drz616khdrjj2-nixos-test-driver-1.1/lib/python3.11/site-packages/test_driver/driver.py", line 147, in run_tests
    self.test_script()
  File "/nix/store/n6zqh28rh2r2h42npr5drz616khdrjj2-nixos-test-driver-1.1/lib/python3.11/site-packages/test_driver/driver.py", line 143, in test_script
    exec(self.tests, symbols, None)
  File "<string>", line 3, in <module>
  File "/nix/store/n6zqh28rh2r2h42npr5drz616khdrjj2-nixos-test-driver-1.1/lib/python3.11/site-packages/test_driver/machine.py", line 614, in execute
    rc = int(self._next_newline_closed_block_from_shell().strip())
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: invalid literal for int() with base 10: 'echo ${PIPESTATUS[0]}'
kill vlan (pid 7)
</pre>
</details>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/1016/74

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2849

SuperSandro2000 commented on nixos/modules/programs/benchexec.nix:
```diff
@@ -0,0 +1,95 @@
+  # nonexistent user before the assertion is evaluated, so
+  # the case of a nonexistent user still does not throw an error,
+  # but instead uses 'nobody'.
+  uid = x: if builtins.isString x then (config.users.users.${x} or config.users.users.nobody).uid else x;
```
Nobody should never be used

lorenzleutgeb replied:
Do you have a suggestion on how to handle this better?

lorenzleutgeb replied:
Fixed by an additional filtering in 8a4cc816e0ff

SuperSandro2000 commented on nixos/modules/programs/benchexec.nix:
```diff
@@ -0,0 +1,95 @@
+    environment.systemPackages = [ cfg.package ];
+
+    systemd = {
+      enableUnifiedCgroupHierarchy = true;
```
This should already be the case, isn't it?

lorenzleutgeb replied:
Converted to an assertion in 8a4cc816e0ff

SuperSandro2000 commented on pkgs/tools/misc/cpu-energy-meter/default.nix:
```diff
@@ -0,0 +1,44 @@
+  };
+
+  patchPhase = ''
+    runHook prePatch
```
patchPhase should always be left as is, use pre and postPatch instead

lorenzleutgeb replied:
Fixed in 2808b9721c4a

SuperSandro2000 commented on pkgs/tools/misc/cpu-energy-meter/default.nix:
```diff
@@ -0,0 +1,44 @@
+  meta = with lib; {
+    description = "A tool for measuring energy consumption of Intel CPUs";
+    homepage = "https://github.com/sosy-lab/cpu-energy-meter";
+    changelog = "https://github.com/sosy-lab/cpu-energy-meter/blob/c528f65/CHANGELOG.md#cpu-energy-meter-12";
```
This link shouldn't be static nor use a short hash

lorenzleutgeb replied:
Fixed in 2808b9721c4a

FliegendeWurst commented on nixos/modules/programs/cpu-energy-meter.nix:
```diff
@@ -0,0 +1,27 @@
+    in
+    lib.mkIf cfg.enable {
+      hardware.cpu.x86.msr.enable = true;
+
```
```suggestion
      boot.kernelModules = [ "cpuid" ];
```
The [readme](https://github.com/sosy-lab/cpu-energy-meter) suggests that cpuid should be loaded I believe.

lorenzleutgeb replied:
It does, but on my machine I could not verify that it is required:

```
$ sudo lsmod | grep -E "^(msr|cpuid)"
msr                    16384  0
$ cpu-energy-meter -d
[DEBUG] GenuineIntel processor found.
[DEBUG] Processor is from family 6 and uses model 0xA0650.
[DEBUG] Using /dev/cpu/0/msr for accessing MSR of socket 0.
[DEBUG] Domain Package is supported.
[DEBUG] Domain Core is supported.
[DEBUG] Domain Uncore is supported.
[DEBUG] Domain DRAM is supported.
[DEBUG] Domain PSYS is supported.
[DEBUG] Measured the following unit multipliers:   RAPL_ENERGY_UNIT=6.103516e-05J   RAPL_DRAM_ENERGY_UNIT=6.103516e-05J
[DEBUG] Not changing UID because not running as root (uid=1000 gid=100).
[DEBUG] Max power consumption of node 0 is 125.000000W.
[DEBUG] Interval time of msr probes set to 1047s, 0ns.
^C
[DEBUG] Received signal 2.
+--------------------------------------+
| CPU Energy Meter            Socket 0 |
+--------------------------------------+
Duration                  7.913466 sec
Package                 169.754578 Joule
Core                    163.473511 Joule
DRAM                     11.148376 Joule
$ sudo modprobe cpuid
$ sudo lsmod | grep -E "^(msr|cpuid)"
cpuid                  16384  0
msr                    16384  0
$ gunzip --stdout /proc/config.gz | grep -s 'CONFIG_X86_CPUID'
CONFIG_X86_CPUID=m
```

@PhilippWendler is loading `cpuid` stricly required?

PhilippWendler replied:
Good question! I think I kept this requirement from the code base from Intel on which cpu-energy-meter was originally based. I don't even know exactly what the cpuid module does. [`cpuid(4)`](https://man7.org/linux/man-pages/man4/cpuid.4.html says it provides `/dev/cpu/*/cpuid`, but we do not use this, we execute the `cpuid` instruction. So as long as there is no implicit requirement somewhere hidden (which I don't think there is), the module should not be necessary.

I will update our readme accordingly.

lorenzleutgeb commented:
I have updated the PR after the release of 23.11. Kindly asking for further review by @h7x4, @SuperSandro2000 or @FliegendeWurst. Thanks!

lorenzleutgeb commented:
Bumped BenchExec to 3.21, fixed merge conflicts, removed `lib.mdDoc`, moved packages to `pkgs/by-name`, added some version tests to `benchexec`. Please review.

SuperSandro2000 commented on nixos/modules/module-list.nix:
```diff
@@ -251,6 +251,7 @@
   ./programs/pantheon-tweaks.nix
   ./programs/partition-manager.nix
   ./programs/plotinus.nix
+  ./programs/pqos-wrapper.nix
```
This should not contain the suffix -wrapper

lorenzleutgeb replied:
Why? You're aware that this is just literally the name of the program/project, right? https://gitlab.com/sosy-lab/software/pqos-wrapper

SuperSandro2000 replied:
Oh, then we should keep it

SuperSandro2000 commented on nixos/modules/programs/pqos-wrapper.nix:
```diff
@@ -0,0 +1,27 @@
+, pkgs
+, ...
+}: {
+  options.programs.pqos-wrapper = {
```
Here we should also drop the wrapper suffix

SuperSandro2000 commented on nixos/modules/programs/pqos-wrapper.nix:
```diff
@@ -0,0 +1,27 @@
+    package = lib.mkPackageOption pkgs "pqos-wrapper" { };
+  };
+
+  config =
+    let
+      cfg = config.programs.pqos-wrapper;
+    in
```
Normally we place this let in straight after the argument list at the top, so before options and config.

lorenzleutgeb replied:
Fixed in 1dd4d2283e09

SuperSandro2000 commented on pkgs/by-name/be/benchexec/package.nix:
```diff
@@ -0,0 +1,62 @@
+    hash = "sha256-bE3brmmLHZQakDKvd47I1hm9Dcsu6DrSeJyjWWtEZWI=";
+  };
+
+  format = "pyproject";
```
This should be pyproject = true

lorenzleutgeb replied:
Fixed in 0e2c44748606

SuperSandro2000 commented on nixos/modules/programs/benchexec.nix:
```diff
@@ -0,0 +1,98 @@
+}:
+let
+  cfg = config.programs.benchexec;
+  opt = options.programs.benchexec;
```
If we use opt just once, we can also hardcode

SuperSandro2000 reviewed (commented):
The module and package should get their own init commits.

lorenzleutgeb committed:
nixos/benchexec: init

lorenzleutgeb committed:
nixos/cpu-energy-meter: init

lorenzleutgeb committed:
nixos/pqos-wrapper: init

lorenzleutgeb commented:
@SuperSandro2000 thanks for the review. I have addressed two of your remarks and split commits so that they do not mix changes to modules with changes to packages. Regarding the "-wrapper" suffix I asked for clarification
